processing/src/main/java/io/druid/query/topn/DimExtractionTopNAlgorithm.java

```diff
@@ -108,19 +109,42 @@ protected void updateResults(
       TopNResultBuilder resultBuilder
   )
   {
-    for (Map.Entry<Comparable, Aggregator[]> entry : aggregatesStore.entrySet()) {
-      Aggregator[] aggs = entry.getValue();
-      if (aggs != null && aggs.length > 0) {
-        Object[] vals = new Object[aggs.length];
-        for (int i = 0; i < aggs.length; i++) {
-          vals[i] = aggs[i].get();
-        }
+    if (needsResultTypeConversion(params)) {
+      final Function<Object, Object> valueTransformer = TopNMapFn.getValueTransformer(
+          query.getDimensionSpec().getOutputType()
+      );
 
-        resultBuilder.addEntry(
-            entry.getKey() == null ? null : entry.getKey().toString(),
-            entry.getKey(),
-            vals
-        );
+      for (Map.Entry<Comparable, Aggregator[]> entry : aggregatesStore.entrySet()) {
+        Aggregator[] aggs = entry.getValue();
+        if (aggs != null && aggs.length > 0) {
+          Object[] vals = new Object[aggs.length];
+          for (int i = 0; i < aggs.length; i++) {
+            vals[i] = aggs[i].get();
+          }
+
+          Comparable convertedKey = (Comparable) valueTransformer.apply(entry.getKey());
+          resultBuilder.addEntry(
+              convertedKey,
+              convertedKey,
+              vals
+          );
+        }
+      }
+    } else {
+      for (Map.Entry<Comparable, Aggregator[]> entry : aggregatesStore.entrySet()) {
+        Aggregator[] aggs = entry.getValue();
+        if (aggs != null && aggs.length > 0) {
+          Object[] vals = new Object[aggs.length];
+          for (int i = 0; i < aggs.length; i++) {
+            vals[i] = aggs[i].get();
+          }
+
+          resultBuilder.addEntry(
+              entry.getKey(),
```

Comment:
The entry won't be null, so I think this is fine. Adding a test with a TopN that uses DimExtraction algorithm with null key values